server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -963,6 +975,20 @@ public boolean isSearchable() {
             return searchable;
         }
 
+        /**
+         * @return whether a shard copy with this role may be the target of a get.
+         */
+        public boolean canHandleGets() {
+            return gets;
+        }
+
+        /**
+         * @return whether a shard copy with this role can be associated with unpromotable shard copies.
```

Comment:
I think this needs a little more precision, because `SEARCH_ONLY` shards are themselves unpromotable and yet return `false` here. I'm not 100% sure of the exact semantics intended here, but I see we're using this to decide whether shards can be relocated via peer recovery or not and I find that confusing. Can we have a separate `canBeRelocatedAsPrimary()` method for that instead? It's a temporary measure, but it'll make it easier to find those places if we're not using one predicate for multiple things.

Reply:
I'm ok with the `canHavePromotables` name but I'd rather have a more precise comment, and also a separate method for the spots where it's used in recoveries.

Reply:
@DaveCTurner I added a new method `canRelocate()`. Please check.